Implement conversion of stablehlo.reduce OP to TTIR dialect

mrakitaTT opened the pull request:
Added MaxOp to TTIR and TTNN dialects and implemented conversion from stablehlo.reduce OP to ttir.sum and ttir.max OPs.

Added necessary tests.

Fixes #510
Fixes #580

mrakitaTT commented on lib/RegisterAll.cpp:
```diff
@@ -37,10 +37,15 @@ void mlir::tt::registerAllDialects(mlir::DialectRegistry &registry) {
+  // Register all dialect conversion passes.
   mlir::tt::registerTTMLIRConversionPasses();
 
+  // Registering -remove-dead-values built-in mlir pass to optimize out the
+  // unused OPs/operands after conversion.
+  mlir::registerPass([]() -> std::unique_ptr<::mlir::Pass> {
+    return mlir::createRemoveDeadValuesPass();
+  });
+
```
For the context, the way `stablehlo.reduce` works is that it is a generic OP which they combine with binary OPs, and it accepts initial value as an argument. For example to do what `ttnn.sum` does, they setup a constant initial value of 0 and then apply reduce with add on tensor by doing `add(initValue, tensor[0])`, then `add(result, tensor[1])`, etc. This is how it looks like:
```
%cst_2 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
%21 = stablehlo.reduce(%20 init: %cst_2) applies stablehlo.add across dimensions = [1] : (tensor<128x10xf32>, tensor<f32>) -> tensor<128xf32>
```

Since TTIR/TTNN SumOp doesn't work with initial value, after we convert `stablehlo.reduce` to `ttir.sum` we end up with unused `stablehlo.constant` OP call. That is where this pass becomes useful, and I think it is also useful in general to have it in our pipeline.

nsmithtt replied:
Can this be written?
```c++
mlir::registerPass(mlir::createRemoveDeadValuesPass);
```

mrakitaTT replied:
Thanks, that works! Not sure why autogenerated pass register functions are always generated with lambda, I copied from there.

nsmithtt reviewed (approved):
Do we have an issue for the runtime changes? It's probably pretty straightforward to add it as part of this change.

mrakitaTT commented:
> Do we have an issue for the runtime changes? It's probably pretty straightforward to add it as part of this change.

@nsmithtt Couldn't reply to signoff comment directly. Good catch, I forgot to add changes in `runtime/lib/ttnn/program.cpp`, will do it now, thanks!

mrakitaTT commented:
I've added TTIR pipelines similar like we have TTNN pipelines, in order to be able to test combination of passes for conversion of input dialects to TTIR. Let me know if someone sees any issues with that.

Concretely right now I had to test combination of `convert-stablehlo-to-ttir` + `remove-dead-values` passes.

mrakitaTT commented:
FYI @jnie-TT I've made similar changes in `runtime/lib/ttnn/program.cpp` for the Reduction OPs like I did in GH-527 for Eltwise OPs.

nobradovictt commented:
> I've added TTIR pipelines similar like we have TTNN pipelines, in order to be able to test combination of passes for conversion of input dialects to TTIR. Let me know if someone sees any issues with that.
> 
> Concretely right now I had to test combination of `convert-stablehlo-to-ttir` + `remove-dead-values` passes.

There is no single test using it? It would be good to have a test targeting it + version with non default version of optional parameter.

sdjordjevicTT commented on lib/Conversion/StableHLOToTTIR/StableHLOToTTIR.cpp:
```diff
@@ -56,9 +61,125 @@ class StableHLOToTTIROpConversionPattern : public OpConversionPattern<SrcOp> {
+  }
+};
+
 struct ConvertStableHLOToTTIRPass
```
Can we move a conversion pass to a separate file? Like we do for TTIRToTTNN, or TTNNToEmitC.

sdjordjevicTT commented on lib/Conversion/StableHLOToTTIR/StableHLOToTTIR.cpp:
```diff
@@ -56,9 +61,125 @@ class StableHLOToTTIROpConversionPattern : public OpConversionPattern<SrcOp> {
     : public ttir::impl::ConvertStableHLOToTTIRBase<
           ConvertStableHLOToTTIRPass> {
+
+  void
```
Can we rewrite these like we do for TTIRToTTNN and EmitC dialect conversions?
populateTTIRToTTNNPatterns, populateTTNNToEmitCPatterns

sdjordjevicTT commented on include/ttmlir/Dialect/TTIR/Pipelines/TTIRPipelines.h:
```diff
@@ -0,0 +1,31 @@
+struct StableHLOToTTIRPipelineOptions
+    : public PassPipelineOptions<StableHLOToTTIRPipelineOptions> {
+  // Option to enable --remove-dead-values optimization pass.
+  Option<bool> removeDeadValuesEnabled{
```
@svuckovicTT, can you take a look at this? We wanted to do this some time ago, but we found a way to remove the ops during the pass's execution, right?

mrakitaTT replied:
I can also remove the op during pass execution, that was the first solution I had in mind, but then I realized that this is gonna require much more complex analysis. If you take a look at example I listed [here](https://github.com/tenstorrent/tt-mlir/pull/585#discussion_r1741263000) there is a constant OP which is used in reduce OP afterwards. I can just remove the constant OP, but then I would have to check not just the reduce OP, but also all the other OPs in that block to see if they use the operand that constant OP produced. That looked like some optimization that MLIR should already have, and indeed I found the `--remove-dead-values` pass.

sdjordjevicTT replied:
Yeah, I see your point, and I agree with you.

mrakitaTT commented:
> > I've added TTIR pipelines similar like we have TTNN pipelines, in order to be able to test combination of passes for conversion of input dialects to TTIR. Let me know if someone sees any issues with that.
> > Concretely right now I had to test combination of `convert-stablehlo-to-ttir` + `remove-dead-values` passes.
> 
> There is no single test using it? It would be good to have a test targeting it + version with non default version of optional parameter.

@nobradovictt Changed all stablehlo tests to use pipeline and added a test for non default version of optional parameter. We don't yet have a concrete case for this optimization until we add support for `stablehlo.constant` OP (check the first comment on this page for more details), so I've added an artificial test for now.

mrakitaTT commented on lib/Conversion/StableHLOToTTIR/StableHLOToTTIR.cpp:
```diff
@@ -56,9 +61,125 @@ class StableHLOToTTIROpConversionPattern : public OpConversionPattern<SrcOp> {
+
+    // Convert based on type.
+    if (srcOpTypeID ==
+        mlir::detail::TypeIDResolver<mlir::stablehlo::AddOp>::resolveTypeID()) {
```
`mlir::cast` throws if types don't match, but I've found that there is a `mlir::isa` function so I've switched to that, thanks!

sdjordjevicTT reviewed (approved):
Great change, thanks!
